// @railgun-community/engine/dist/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyED25519 = exports.getPublicViewingKey = exports.convertTransactionStructToCommitmentSummary = exports.toUTF8String = exports.fromUTF8String = exports.ByteUtils = exports.ByteLength = void 0;
// Note: we purposefully do not export everything, in order to reduce the number of public APIs
var bytes_1 = require("./bytes");
Object.defineProperty(exports, "ByteLength", { enumerable: true, get: function () { return bytes_1.ByteLength; } });
Object.defineProperty(exports, "ByteUtils", { enumerable: true, get: function () { return bytes_1.ByteUtils; } });
Object.defineProperty(exports, "fromUTF8String", { enumerable: true, get: function () { return bytes_1.fromUTF8String; } });
Object.defineProperty(exports, "toUTF8String", { enumerable: true, get: function () { return bytes_1.toUTF8String; } });
var commitment_1 = require("./commitment");
Object.defineProperty(exports, "convertTransactionStructToCommitmentSummary", { enumerable: true, get: function () { return commitment_1.convertTransactionStructToCommitmentSummary; } });
__exportStar(require("./ecies"), exports);
var keys_utils_1 = require("./keys-utils");
Object.defineProperty(exports, "getPublicViewingKey", { enumerable: true, get: function () { return keys_utils_1.getPublicViewingKey; } });
Object.defineProperty(exports, "verifyED25519", { enumerable: true, get: function () { return keys_utils_1.verifyED25519; } });
